### PR TITLE
Support Amazon Linux and Xenial

### PR DESCRIPTION
**Pull request by patrobinson**
`--wait-for-rc` forces ami-spec to wait for the system bootstrap to finish before running tests. This is helpful as `sshd` can start quite early and our tests might execute before the system is fully booted, giving false failures.

But support was limited to Ubuntu trusty. We also failed with obscure or no error messages if it took no effect.

This PR works to support more OS types and print out explicit warnings when it has no impact.
It also introduces some integration tests. Due to the required security permissions for Xenial those can't be run in Travis, so must be run locally.

### CODE REVIEW

File: spec/containers/Dockerfile.amazon_linux [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM amazonlinux:1
```
Comment (viraptor) [inline spec/containers/Dockerfile.amazon_linux]: Not sure what to do about this, but just looking at the repository I'd find the files here confusing. Maybe some README in `spec/containers` about what you're trying to do would help? Or a specific subdirectory like `spec/containers/wait_for_rc_environments/...` ?

File: lib/ami_spec/wait_for_rc.rb [hunk 2]
```diff
@@ -2,15 +2,39 @@
...
-        # Wait for SystemV to start
-        # This only works for Ubuntu with upstart.
-        # Detecting OS and Release will need something like this
-        # https://github.com/mizzy/specinfra/blob/master/lib/specinfra/helper/detect_os/debian.rb
-        ssh.exec 'while /usr/sbin/service rc status | grep -q "^rc start/running, process"; do sleep 1; done'
+    def self.wait(ip_address, user, key, port=22)
+      Net::SSH.start(ip_address, user, keys: [key], paranoid: false, port: port) do |ssh|
+        distrib_stdout = ""
+        # Determine the OS family
+        ssh.exec!("source /etc/*release && echo -n $DISTRIB_ID && echo -n $ID") do |channel, stream, data|
```
[the remark this answers is not in the document]
Comment (patrobinson) [reply lib/ami_spec/wait_for_rc.rb]: The README kinda hints what they're there for, but I'll make it more explicit.

File: spec/containers/Dockerfile.amazon_linux (hunk 1 above)
Comment (patrobinson) [reply spec/containers/Dockerfile.amazon_linux]: The README kinda hints what they're there for, but I'll make it more explicit.

File: lib/ami_spec/wait_for_rc.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrobinson) [reply lib/ami_spec/wait_for_rc.rb]: Amazon Linux doesn't have `lsb_release` installed

[the remark this answers is not in the document]
Comment (viraptor) [reply lib/ami_spec/wait_for_rc.rb]: Grrrrr....

[the remark this answers is not in the document]
Comment (patrobinson) [reply lib/ami_spec/wait_for_rc.rb]: If it helps Puppet Facter uses these files (Which has been re-writtin in C++ 😱) https://github.com/puppetlabs/facter/blob/78fbdcf7981e065c289e192903fc44b52da33799/lib/inc/internal/facts/linux/release_file.hpp#L109

[the remark this answers is not in the document]
Comment (viraptor) [reply lib/ami_spec/wait_for_rc.rb]: 😱😱😱

File: spec/containers/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+version: '3'
+services:
+  xenial:
+    build:
+      context: .
+      dockerfile: Dockerfile.xenial
+    ports:
+      - "1122:22"
+    # --security-opt seccomp=unconfined --tmpfs /run --tmpfs /run/lock -v /sys/fs/cgroup:/sys/fs/cgroup:ro
```
Comment (IcyWednesdays) [inline spec/containers/docker-compose.yml]: Do you still need this comment?

Comment (patrobinson) [reply spec/containers/docker-compose.yml]: It's just a handy hint if you want to run the container from the command line

File: spec/wait_for_rc_spec.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    end
+  end
+
+  context 'amazon linux server' do
+    let(:ssh_port) { 1124 }
+    it 'executes without printing any errors' do
+      expect { described_class.wait("localhost", "root", private_key_file, ssh_port) }.to_not output.to_stdout
+    end
+  end
+end
```
Comment (patrobinson) [inline spec/wait_for_rc_spec.rb]: We're not testing that we actually _wait_ for RC to "stop", we're just testing that no error messages get thrown. I put that in the too hard basket.

File: spec/containers/docker-compose.yml (hunk 3 above)
Comment (IcyWednesdays) [reply spec/containers/docker-compose.yml]: Got it 👍

**Pull request merged**